<launch>
  <param name="whitelist_topics"
         value="$(arg name)/amcl_pose,$(arg name)/laser_odom/wsubscan,$(arg name)/move_server/cancel,$(arg name)/move_server/goal,$(arg name)/move_server/results,$(arg name)/move_server/status,$(arg name)/move_server/feedback,$(arg name)/laser_odom/lodom,/measurements,$(arg name)/path"/>
  <param name="whitelist_services" value=""/>
  <param name="whitelist_params" value=""/>
  <param name="blacklist_topics" value="/tf,/rosout" />

  <group ns="$(arg name)">
    <param name="tf_prefix" value="$(arg name)"/>

    <!--============================= Control =============================-->
    <!-- These parameters work well for a scarab with the big pulleys/belts -->
    <node pkg="acroname_moto" type="acroname_moto_node" name="motor"
          output="screen" respawn="false">
      <param name="portname" value="acroname"/>
      <param name="frameid" value="odom" />
      
      <param name="axle_width" value="0.23"/>
      <param name="max_wheel_vel" value="10.0"/>
      <param name="min_wheel_vel" value="0.0"/>
      <param name="wheel_diam" value="0.1"/>
      <param name="encoder_count_per_motor_rev" value="64" />

      <param name="left_channel" value="0"/>
      <param name="right_channel" value="1"/>

      <param name="motor_to_wheel_ratio" value="80.5"/>
      <param name="pid_param_p" value="4.5"/>
      <param name="pid_param_i" value="0.5"/>
      <param name="pid_param_d" value="2.3"/>
      <param name="pid_period" value="0.02" />

      <param name="pwm_freq" value="40000" />
      <param name="freq" value="15" />
      <param name="baud" value="38400" />
    </node>

    <!-- =========================== Localization ============================ -->
    <node name="laser" pkg="urg_node" type="urg_node" respawn="false"
          output="screen">
      <param name="laser_transform/x" value="0.0" />
      <param name="laser_transform/y" value="0.0" />
      <param name="laser_transform/z" value="0.198" />
      <param name="rate" value="10" />
      <param name="baud" value="19200" />
    </node>

    <node name="laser_subsample" pkg="laser_subsample" type="laser_subsample"
          output="screen">
      <remap from="~scan" to="laser_odom/wscan" />
      <remap from="~subscan" to="laser_odom/wsubscan" />
    </node>

    <node pkg="map_server" name="map" type="map_server" output="screen"
          args="$(find upenn_maps)/maps/levine_towne_map.yaml"/>

    <node pkg="laser_odom" name="laser_odom" type="laser_odom" output="screen">
      <remap from="~scan" to="laser/scan"/>
      <remap from="~odom" to="motor/odom"/>
      <param name="map_frame" value="/map" type="string" />
      <param name="lodom_frame" value="$(arg name)/lodom" />
      <param name="wodom_frame" value="$(arg name)/wodom" />
      <param name="lbase_frame" value="$(arg name)/lbase" />
      <param name="wbase_frame" value="$(arg name)/wbase" />
      <param name="llaser_frame" value="$(arg name)/llaser" />
      <param name="wlaser_frame" value="$(arg name)/wlaser" />
      <param name="map_frame" value="map" type="string" />

      <param name="freq" value="8" />
    </node>

    <node pkg="pamcl" name="pamcl" type="amcl" respawn="false" output="screen">
      <!-- These params are typically overriden in launch file -->
      <param name="initial_pose_x" value="$(arg x)" type="double"/>
      <param name="initial_pose_y" value="$(arg y)" type="double"/>
      <param name="initial_pose_a" value="$(arg a)" type="double"/>
      
      <param name="global_frame_id" value="/map" />

      <param name="initial_pose_xx" value="0.4" type="double"/>
      <param name="initial_pose_yy" value="0.4" type="double"/>
      <param name="initial_pose_aa" value="0.2" type="double"/>

      <remap from="scan" to="laser_odom/wscan"/>
      <remap from="static_map" to="static_map"/>
      <remap from="lodom" to="laser_odom/lodom"/>

      <param name="odom_frame_id" value="wodom"/>
      <param name="base_frame_id" value="wbase"/>
      <param name="global_frame_id" value="map"/>

      <param name="odom_model_type" value="diff"/>
      <param name="transform_tolerance" value="1.0" />
      <param name="laser_max_beams" value="150"/>
      <param name="min_particles" value="300"/>
      <param name="max_particles" value="600"/>
      <param name="kld_err" value="0.1"/>
      <param name="kld_z" value="0.96"/>

      <param name="odom_alpha1" value="0.5"/>
      <param name="odom_alpha2" value="0.5"/>
      <param name="odom_alpha3" value="0.5"/>
      <param name="odom_alpha4" value="0.5"/>

      <param name="laser_z_hit" value="0.90"/>
      <param name="laser_z_rand" value="0.10"/>
      <param name="laser_sigma_hit" value="0.1"/>
      <param name="laser_lambda_short" value="0.1"/>
      <param name="laser_model_type" value="likelihood_field"/>
      <param name="laser_likelihood_max_dist" value="1.5"/>

      <param name="update_min_d" value="0.1"/>
      <param name="update_min_a" value="0.1"/>

      <param name="resample_interval" value="1"/>
      <param name="recovery_alpha_slow" value="0.0"/>
      <param name="recovery_alpha_fast" value="0.0"/>
    </node>

    <!--=========================== Navigation ============================-->
    <node pkg="rvo_move" type="move_server" name="move_server" output="screen">
      <rosparam file="$(find rvo_move)/config/scarab_move.yaml" command="load" />
      <remap from="cmd_vel" to="motor/cmd_vel" />
      <remap from="map" to="/nav_static_map" />
      <remap from="range_map" to="/range_static_map" />

      <param name="nbots" value="3" />
      <param name="id" value="$(arg id)" />
      <param name="robot0" value="/scarab26" />
      <param name="robot1" value="/scarab27" />
      <param name="robot2" value="/scarab32" />

      <param name="obstacle_file" value="$(find source_estimation)/maps/polygons.txt" />
    </node>
  </group>
</launch>
